fluss-server/src/main/java/org/apache/fluss/server/metadata/CoordinatorMetadataFunctionProvider.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.fluss.server.metadata;
+
+import org.apache.fluss.metadata.PhysicalTablePath;
+import org.apache.fluss.metadata.TableBucket;
+import org.apache.fluss.metadata.TableInfo;
+import org.apache.fluss.metadata.TablePartition;
+import org.apache.fluss.metadata.TablePath;
+import org.apache.fluss.server.RpcServiceBase;
+import org.apache.fluss.server.coordinator.CoordinatorContext;
+import org.apache.fluss.server.coordinator.MetadataManager;
+import org.apache.fluss.server.zk.ZooKeeperClient;
+import org.apache.fluss.server.zk.data.LeaderAndIsr;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/** The coordinator metadata function provider. */
+public class CoordinatorMetadataFunctionProvider implements MetadataFunctionProvider {
+
+    private final ZooKeeperClient zkClient;
+
+    private final CoordinatorMetadataCache metadataCache;
+
+    private final CoordinatorContext ctx;
+
+    private final MetadataManager metadataManager;
+
+    public CoordinatorMetadataFunctionProvider(
+            ZooKeeperClient zkClient,
+            CoordinatorMetadataCache metadataCache,
+            CoordinatorContext ctx,
+            MetadataManager metadataManager) {
+        this.zkClient = zkClient;
+        this.metadataCache = metadataCache;
+        this.ctx = ctx;
+        this.metadataManager = metadataManager;
+    }
+
+    @Override
+    public Optional<TableMetadata> getTableMetadataFromCache(TablePath tablePath) {
+        TableInfo tableInfo = metadataManager.getTable(tablePath);
```

Comment:
1. we can move this after the `if` condition, to reduce the cost of zk operation. 
2. this is still reading from zk instead of from cache, I think we can get TableInfo from `ctx` as well